amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ActivityImporter.java

```diff
@@ -146,77 +155,92 @@ private void init(Map<String, Object> newJson, boolean update, String endpointCo
             return this;
         }
 
-        checkPermissions(update, ampActivityId, modifiedBy);
+        checkPermissions(update, activityId, modifiedBy);
 
         if (!errors.isEmpty()) {
             return this;
         }
 
         // check if any error were already detected in upper layers
-        Map<Integer, ApiErrorMessage> existingErrors = (TreeMap<Integer, ApiErrorMessage>) newJson.get(ActivityEPConstants.INVALID);
+        Map<Integer, ApiErrorMessage> existingErrors = (TreeMap<Integer, ApiErrorMessage>)
+                newJson.get(ActivityEPConstants.INVALID);
 
         if (existingErrors != null && existingErrors.size() > 0) {
             errors.putAll(existingErrors);
         }
 
-        if (ampActivityId != null) {
-            try {
-                oldActivity = loadActivity(ampActivityId);
-            } catch (DgException e) {
-                logger.error(e.getMessage());
-                errors.put(ActivityErrors.ACTIVITY_NOT_LOADED.id, ActivityErrors.ACTIVITY_NOT_LOADED);
+        try {
+            activityService.doWithLock(activityId, modifiedBy.getAmpTeamMemId(),
+                    () -> importOrUpdateActivity(activityId));
+        } catch (Throwable e) {
+            // error is not always logged at source; better duplicate it than have none
+            logger.error("Import failed", e);
+            if (e instanceof ActivityLockNotGrantedException) {
+                logger.error("Cannot aquire lock during IATI update for activity " + activityId);
+                Long userId = ((ActivityLockNotGrantedException) e).getUserId();
+                String memberName = TeamMemberUtil.getTeamMember(userId).getMemberName();
+                errors.put(ActivityErrors.ACTIVITY_IS_BEING_EDITED.id,
+                        ActivityErrors.ACTIVITY_IS_BEING_EDITED.withDetails(memberName));
+            } else if (e instanceof StaleStateException) {
+                throw new RuntimeException("Activity updated in meantime without using gatekeeper.", e);
+            } else if (errors.isEmpty()) {
+                throw new RuntimeException(e);
+            } else if (!(e instanceof ImportFailedException)) {
+                addError(new ApiExceptionMapper().getApiErrorMessageFromException(e));
             }
+        } finally {
+            ActivityValidationContext.set(null);
         }
+    
+        updateResponse(update);
 
-        sanityChecks();
-
-        String activityId = ampActivityId == null ? null : ampActivityId.toString();
-        String key = null;
-
-        Long currentVersion = null;
-
+        return this;
+    }
+    
+    /**
+     * Import or updates activity.
+     * @param activityId
+     */
+    private void importOrUpdateActivity(Long activityId) {
+        boolean oldActivityDraft = false;
         try {
-            // initialize new activity
-            InterchangeUtils.getSessionWithPendingChanges();
```

Comment:
getSessionWithPendingChanges is no longer called

Reply:
The flush mode will be set in doWithLock method